README.md
<p align="center">
  <a href="https://github.com/Fliggy-Mobile">
    <img width="200" src="https://gw.alicdn.com/tfs/TB1a288sxD1gK0jSZFKXXcJrVXa-360-360.png">
  </a>
</p>

<h1 align="center">fradio</h1>


<div align="center">

<p>A radio component suitable for almost any radio scenario.</p>

<p>Supports excellent interactive special effects, as well as a simple multi-interactive state view switching construction like eating a hamburger 🍔. You will fall in love with TA ❤.️</p>

<p><strong>Author：<a href="https://github.com/chenBingX">Newton</a>(<a href="coorchice.cb@alibaba-inc.com">coorchice.cb@alibaba-inc.com</a>)</strong></p>

<p>

<a href="https://pub.dev/packages/fradio#-readme-tab-">
    <img height="20" src="https://img.shields.io/badge/Version-1.0.1-important.svg">
</a>


<a href="https://github.com/Fliggy-Mobile/fradio">
    <img height="20" src="https://img.shields.io/badge/Build-passing-brightgreen.svg">
</a>


<a href="https://github.com/Fliggy-Mobile">
    <img height="20" src="https://img.shields.io/badge/Team-FAT-ffc900.svg">
</a>

<a href="https://www.dartcn.com/">
    <img height="20" src="https://img.shields.io/badge/Language-Dart-blue.svg">
</a>

<a href="https://pub.dev/documentation/fradio/latest/fradio/fradio-library.html">
    <img height="20" src="https://img.shields.io/badge/API-done-yellowgreen.svg">
</a>

<a href="http://www.apache.org/licenses/LICENSE-2.0.txt">
   <img height="20" src="https://img.shields.io/badge/License-Apache--2.0-blueviolet.svg">
</a>

<p>
<p>

<img height="500" src="https://raw.githubusercontent.com/chenBingX/img/master/Flutter/fradio/fradio_all.png">

</div>

**English | [简体中文](https://github.com/Fliggy-Mobile/fradio/blob/master/README_CN.md)**

> Like it? Please cast your **Star**  🥰 ！

# ✨ Features

- Wonderful interactive animation

- Support precise control of rounded corners

- Wonderful gradient effect support

- Simple but effective multi-state view building support

- Unimaginable flexible configurable items

# 🛠 Guide

## ⚙️ Parameters


### 🔩 Basic parameters

|Param|Type|Necessary|Default|desc|
|---|---|:---:|---|---|
|value|`<T>`|false|null|The value represented by [FRadio]. When [groupValue] == [value], enter the selected state.|
|groupValue|`<T>`|false|null|The currently selected value of the radio group. When [groupValue] == [value], enter the selected state.|
|onChanged|`ValueChanged<T>`|false|null|Callback when [FRadio] is selected|
|enable|bool|false|true|it's usable or not. Unavailable [FRadio] will not be able to change the current state by clicking. Through [disableNormal] and [disableSelected], you can customize the style in the unavailable state.|
|toggleable|bool|false|false|Is it possible to uncheck it.|
|width|double|false|27|width|
|height|double|false|27|height|
|normal|Widget|false|null|Unselected state style|
|selected|Widget|false|null|Selected state style|
|disableNormal|Widget|false|null|Unavailable style when unchecked|
|disableSelected|Widget|false|null|Unavailable styles selected|
|hover|Widget|false|null|The style when the mouse enters|
|focusNode|FocusNode|false|null|focus|
|autofocus|bool|false|false|Whether to allow automatic focus|


### 🔩 Default constructor extension parameters

|Param|Type|Necessary|Default|desc|
|---|---|:---:|---|---|
|normalColor|Color|false|`Color(0xffd9d9d9)`|Unselected color|
|selectedColor|Color|false|`Color(0xff2593fc)`|Selected color|
|hasSpace|bool|false|true|Is there a gap between the inner padding and the edges? The default is true. The spacing is provided by [FRadio] the golden ratio, developers do not need to care.|
|border|double|false|null|The border is wide. By default, [FRadio] provides the golden ratio, and developers do not need to care.|
|child|Widget|false|null|The decoration components in the unselected state are at the top level.|
|selectedChild|Widget|false|null|The decorative component in the selected state is at the top level.|
|hoverChild|Widget|false|null|The decoration component when hovering the mouse is at the top level.|
|gradient|Gradient|false|null|This property allows to configure the gradient effect in the selected state, which will override [selectedColor].|
|duration|Duration|false|`Duration(milliseconds: 150)`|Duration of state switching animation.|
|fill|bool|false|true|When selected, whether to allow internal filling.|
|corner|FRadioCorner|false|Circle|Corner.|

## 📺 Demo

### 🔩 Basic Demo


![](https://raw.githubusercontent.com/chenBingX/img/master/Flutter/fradio/fradio_demo1.gif)

```dart
FRadio(
  value: 1,
  groupValue: groupValue_1,
  onChanged: (value) {
    setState(() {
      groupValue_1 = value;
    });
  },
),

FRadio(
  value: 2,
  groupValue: groupValue_1,
  onChanged: (value) {
    setState(() {
      groupValue_1 = value;
    });
  },
)
```

A regular **FRadio** is very simple to build. You only need to configure:

`value`: It will indicate the current value represented by **FRadio**.

`groupValue`：Indicates the selected value of the current radio group.

`onChanged`：When selected, it will callback. In this callback, you need to assign the value `value` of **FRadio** to the radio group` groupValue` to complete the switch of group options.

> 💡It is worth noting that **FRadio** will only be selected when `groupValue` ==` value`. This is very flexible, that is to say, you only need to change the groupValue to switch options at will.

### 👀 Enable & Toggleable

![](https://raw.githubusercontent.com/chenBingX/img/master/Flutter/fradio/fradio_demo2.gif)

```dart
FRadio(
  value: 1,
  groupValue: groupValue_2,
  onChanged: (value) {
    setState(() {
      groupValue_2 = value;
    });
  },
  toggleable: true,
  enable: enable_1,
  selectedColor: Color(0xffffc900),
),
```

Through the `enable` attribute, you can control the availability of **FRadio**. When **FRadio** is not available, the unavailable style will be displayed. They of course include **selected** and **unselected**.

The unavailable styles of **FRadio** built by the default constructor are grayed out.

If you want to build different unavailable styles, you can configure `disableNormal` (**unchecked unavailable styles**) and` disableSelected` (**check unavailable styles**) through the `FRadio.custom ()` constructor.

**FRadio** will allow the user to cancel the selection, which can be achieved by configuring `toggleable: true`. Of course, by default, we do not open this feature.

### 🔆 Space & Corner & Border

![](https://raw.githubusercontent.com/chenBingX/img/master/Flutter/fradio/fradio_demo3.gif)

```dart
FRadio(
  width: 100,
  height: 50,
  value: 1,
  groupValue: groupValue_3,
  onChanged: (value) {
    setState(() {
      groupValue_3 = value;
    });
  },
  selectedColor: Color(0xffffc900),
  corner: FRadioCorner(leftTopCorner: 6, leftBottomCorner: 6),
  border: 1,
  hasSpace: false,
  selectedChild:
      Text("FSuper", style: TextStyle(color: Color(0xff333333))),
  child: Text("FSuper", style: TextStyle(color: Color(0xff333333))),
)
```

In this chestnut 🌰, **FRadio** shows several abilities that are not possible for radio components.

Through the `corner` attribute, you can accurately control the corners so that **FRadio** is no longer a circle.

`hasSpace` gives you the freedom to choose whether you want to keep the distance between the inner padding and the edges, as it is by default.

Usually, **FRadio** has a set of exquisite calculation methods internally to provide a very beautiful view. By default, **FRadio** will automatically calculate the width of the border to maintain a harmonious aesthetic.

But you can still configure it as needed through the `border` attribute.

### 🌈 Gradient

![](https://raw.githubusercontent.com/chenBingX/img/master/Flutter/fradio/fradio_demo4.gif)

```dart
FRadio(
  width: 50,
  height: 50,
  value: index + 1,
  groupValue: groupValue_4,
  onChanged: (value) {
    setState(() {
      groupValue_4 = value;
    });
  },
  gradient: LinearGradient(
    colors: [
      Color(0xffFEFDBB),
      Color(0xffFFE16C),
      Color(0xffEA9D1C),
      Color(0xffD46307),
    ],
    begin: Alignment(-0.1, -0.9),
    end: Alignment(1.0, 1.0),
    stops: [0.0, 0.2, 0.7, 1.0],
  ),
  selectedColor: Color(0xffffc900),
  hasSpace: false,
  border: 1.5,
  child: Text(
    "\$${5 * (index + 1)}",
    style: TextStyle(color: Color(0xffd9d9d9), fontSize: 13),
  ),
  hoverChild: Text(
    "\$${5 * (index + 1)}",
    style:
        TextStyle(color: Colors.deepOrangeAccent, fontSize: 13),
  ),
  selectedChild: Text("\$${5 * (index + 1)}",
      style: TextStyle(
          color: Colors.deepOrangeAccent, fontSize: 13)),
)
```

You see, this is the magical use of gradients.

`gradient` accepts multiple types of gradient configurations. If you plan to make the radio view look more colorful or more spatial, you can try it.


### 🍭 Decoration

![](https://raw.githubusercontent.com/chenBingX/img/master/Flutter/fradio/fradio_demo5.gif)

```dart
FRadio(
  width: 80,
  height: 80,
  value: 1,
  groupValue: groupValue_5,
  onChanged: (value) {
    setState(() {
      groupValue_5 = value;
    });
  },
  child: Image.asset("assets/emoji_0.png", width: 50),
  hoverChild: Image.asset("assets/emoji_1.png", width: 50),
  selectedChild: Image.asset("assets/emoji_2.png", width: 50),
  hasSpace: false,
  toggleable: true,
  selectedColor: Color(0xffffc900),
  border: 1.5,
)
```

**FRadio** magically makes the entire radio no longer boring!

`child` and` selectedChild` configure the top-level decorative components.

How is the mouse hover effect constructed?

The border color will automatically accept the selected color configured by `selectedColor` when hovering. The decoration can be configured through `hoverChild`.

When `hoverChild` is not configured, the selected state decoration configured by` selectedChild` will be accepted.

how about it? This is so interesting!

### 🎨 This is customizable!

![](https://raw.githubusercontent.com/chenBingX/img/master/Flutter/fradio/fradio_demo6.gif)

If the default constructor does not satisfy you, what else do you need?

By the way, **Customize**!

Through the `FRadio.custom ()` constructor, you can configure the style of **FRadio** in different interactive states through: `normal`,` selected`, `disableNormal`,` disableSelected`, `hover`.

This will untie the hemp rope in your hand, you are free, and you can start to create at will.

![](https://raw.githubusercontent.com/chenBingX/img/master/Flutter/fradio/fradio_demo7.gif)

```dart
ListView.builder(
  itemCount: list.length,
  itemBuilder: (context, index) {
    return FRadio.custom(
      value: index,
      groupValue: groupValue_7,
      onChanged: (value) {
        setState(() {
          groupValue_7 = value;
        });
      },
      normal: Container(
        width: 250,
        height: 100,
        color: Colors.white,
        padding: EdgeInsets.only(left: 12, right: 12),
        child: text(index),
      ),
      hover: Container(
        width: 250,
        height: 100,
        color: Color(0xff212121).withOpacity(0.05),
        padding: EdgeInsets.only(left: 12, right: 12),
        child: text(index),
      ),
      selected: Container(
        padding: EdgeInsets.only(left: 12, right: 12),
        color: Color(0xff607D8B).withOpacity(0.2),
        child: Row(
          crossAxisAlignment: CrossAxisAlignment.center,
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Text("${list[index]}"),
            Icon(
              Icons.check,
              size: 18,
            ),
          ],
        ),
      ),
    );
  },
)
```

Now you are free, so don't skimp on creativity.

**FRadio** is not just a small round button, it can solve almost all radio problems.




# 😃 How to use？

Add dependencies in the project `pubspec.yaml` file:

## 🌐 pub dependency

```
dependencies:
  fradio: ^<version number>
```

> ⚠️ Attention，please go to [**pub**] (https://pub.dev/packages/fradio) to get the latest version number of **FRadio**

## 🖥 git dependencies

```
dependencies:
  fradio:
    git:
      url: 'git@github.com:Fliggy-Mobile/fradio.git'
      ref: ''<Branch number or tag number>'
```

> ⚠️ Attention，please refer to [**FRadio**] (https://github.com/Fliggy-Mobile/fradio) official project for branch number or tag.


# 💡 License

```
Copyright 2020-present Fliggy Android Team <alitrip_android@list.alibaba-inc.com>.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at following link.

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

```


### Like it? Please cast your [**Star**](https://github.com/Fliggy-Mobile/fradio)  🥰 ！



--

# How to run Demo project?

1. **clone** project to local

2. Enter the project `example` directory and run the following command

```
flutter create .
```

3. Run the demo in `example`


CHANGELOG.md
## 1.1.0
- 适配空安全

## 1.0.0
- The first official version

## 1.0.1

- remove redundant code

- update pub info


test/fradio_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:fradio/fradio.dart';

void main() {

}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:fradio_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


android/src/main/java/com/taobao/fapi/fradio/FradioPlugin.java
package com.taobao.fapi.fradio;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** FradioPlugin */
public class FradioPlugin implements MethodCallHandler {
  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "fradio");
    channel.setMethodCallHandler(new FradioPlugin());
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }
}


example/lib/main.dart
import 'dart:math';

import 'package:fbutton/fbutton.dart';
import 'package:flutter/material.dart';
import 'package:fradio/fradio.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: FRadioPage(),
    );
  }
}

class FRadioPage extends StatefulWidget {
  @override
  _FRadioPageState createState() => _FRadioPageState();
}

class _FRadioPageState extends State<FRadioPage> {
  int groupValue_1 = 0;
  bool enable_1 = true;

  int groupValue_2 = 0;

  int groupValue_3 = 0;

  int groupValue_4 = 0;

  int groupValue_5 = 0;

  int groupValue_6 = 0;

  int groupValue_7 = -1;

  List<String> list;

  @override
  void initState() {
    super.initState();
    list = [];
    list.add("Kicks");
    list.add("No Time To Die");
    list.add("Pavane in F-Sharp Minor");
    list.add("Get Stupid");
    list.add("AAA");
    list.add("Hostage");
    list.add("Boundaries");
    list.add("Broken");
    list.add("Lo Que Siento");
    list.add("Bored");
    list.add("Savage");
    list.add("Paris");
    list.add("Bad guy");
    list.add("My boy");
    list.add("I love you");
    list.add("Copycat");
    list.add("Lovely");
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Color(0xffffc900),
        title: const Text('FRadio'),
        centerTitle: true,
      ),
      body: SingleChildScrollView(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: <Widget>[
            buildTitle("FRadio"),
            buildSmallMargin(),

            /// 基础用法
            baseDemo(),
            buildSmallMargin(),
            buildDesc('Creating a group of FRadio is very simple.'),
            buildMiddleMargin(),
            buildTitle("Enable & Toggleable"),
            buildSmallMargin(),

            /// Enable & Toggleable
            enableToggleableDemo(),
            buildSmallMargin(),
            FButton(
              width: 100,
              corner: FButtonCorner.all(6),
              text: "Toggleable",
              textColor: Colors.white,
              fontSize: 12,
              onPressed: () {
                setState(() {
                  enable_1 = !enable_1;
                });
              },
              color: Color(0xffffc900),
              shadowColor: Colors.black.withOpacity(0.2),
              shadowBlur: 1.0,
              effect: true,
            ),
            buildMiddleMargin(),
            buildTitle("Space & Corner & Border"),
            buildMiddleMargin(),

            /// Space & Corner
            spaceCornerDemo(),
            buildMiddleMargin(),
            buildTitle("Gradient"),
            buildMiddleMargin(),

            /// gradient
            gradientDemo(),

            buildMiddleMargin(),
            buildTitle("Child"),
            buildMiddleMargin(),

            /// Child
            childDemo(),

            buildMiddleMargin(),
            buildTitle("Custom 1"),

            /// Custom 1
            customDemo1(),
//            Transform.scale(scale: 1.0, child: customDemo(),),

            buildMiddleMargin(),
            buildTitle("Custom 2"),
            buildMiddleMargin(),

            /// Custom 2
            customDemo2(),
            buildMiddleMargin(),
            buildMiddleMargin(),
            buildMiddleMargin(),
            buildMiddleMargin(),
          ],
        ),
      ),
    );
  }

  Container customDemo2() {
    return Container(
      width: 250,
      height: 400,
      decoration: BoxDecoration(color: Colors.white, boxShadow: [
        BoxShadow(
          color: Color(0xffBDBDBD),
          blurRadius: 5.0,
          offset: Offset(2.0, 2.0),
        )
      ]),
      child: ListView.builder(
        shrinkWrap: true,
        itemCount: list.length,
        physics: BouncingScrollPhysics(),
        itemBuilder: (context, index) {
          return FRadio.custom(
            value: index,
            groupValue: groupValue_7,
            onChanged: (value) {
              setState(() {
                groupValue_7 = value;
              });
            },
            normal: Container(
              width: 250,
              height: 100,
              color: Colors.white,
              padding: EdgeInsets.only(left: 12, right: 12),
              child: text(index),
            ),
            hover: Container(
              width: 250,
              height: 100,
              color: Color(0xff212121).withOpacity(0.05),
              padding: EdgeInsets.only(left: 12, right: 12),
              child: text(index),
            ),
            selected: Container(
              padding: EdgeInsets.only(left: 12, right: 12),
              color: Color(0xff607D8B).withOpacity(0.2),
              child: Row(
                crossAxisAlignment: CrossAxisAlignment.center,
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  Text("${list[index]}"),
                  Icon(
                    Icons.check,
                    size: 18,
                  ),
                ],
              ),
            ),
          );
        },
      ),
    );
  }

  Row text(int index) {
    return Row(
      crossAxisAlignment: CrossAxisAlignment.center,
      children: [
        Text("${list[index]}"),
      ],
    );
  }

  Widget customDemo1() {
    int count = 3;
    return Stack(
      children: [
        Image.asset("assets/icon_radio.png"),
        Positioned(
          top: 230,
          left: 150,
          child: SizedBox(
            height: 30,
            child: ListView.builder(
                scrollDirection: Axis.horizontal,
                itemCount: count,
                shrinkWrap: true,
                itemBuilder: (context, index) {
                  return Padding(
                    padding: EdgeInsets.only(right: index == count - 1 ? 0 : 1),
                    child: FRadio.custom(
                      width: 18,
                      height: 30,
                      value: index + 1,
                      groupValue: groupValue_6,
                      onChanged: (value) {
                        setState(() {
                          groupValue_6 = value;
                        });
                      },
                      toggleable: true,
                      normal: demo3d(Color(0xff686868)),
                      hover: demo3d(Color(0xff9e9e9e)),
                      selected: Transform.scale(
                        scale: 0.85,
                        alignment: Alignment.center,
                        child: Container(
                          child: demo3d(Color(0xffD84315)),
                          decoration: BoxDecoration(
                            boxShadow: [
                              BoxShadow(
                                color: Color(0xffFFAB91),
                                blurRadius: 2.0,
                              ),
                            ],
                          ),
                        ),
                      ),
                    ),
                  );
                }),
          ),
        ),
      ],
    );
  }

  Widget demo3d(Color color) {
    return Stack(
      children: [
        /// top
        Positioned(
          top: 0,
          left: 3,
          child: Transform(
            transform: Matrix4.skewX(-pi / 6),
            child: Container(
              width: 15,
              height: 5,
              decoration: BoxDecoration(
                borderRadius: BorderRadius.all(Radius.circular(0)),
                gradient: LinearGradient(
                    colors: [color.withOpacity(0.8), color.withOpacity(0.3)],
                    begin: Alignment.bottomCenter,
                    end: Alignment.topCenter),
              ),
            ),
          ),
        ),

        /// positive
        Positioned(
          top: 5,
          child: Container(
              width: 15,
              height: 25,
              decoration: BoxDecoration(
                color: color,
                borderRadius: BorderRadius.all(Radius.circular(0)),
              )),
        ),

        /// side
        Positioned(
          top: 5,
          left: 15,
          child: Transform(
            transform: Matrix4.skewY(-pi / 3),
            child: Container(
                width: 3,
                height: 25,
                decoration: BoxDecoration(
                    borderRadius: BorderRadius.all(Radius.circular(0)),
                    gradient: LinearGradient(
                        colors: [
                          color.withOpacity(0.8),
                          color.withOpacity(0.5)
                        ],
                        begin: Alignment.centerLeft,
                        end: Alignment.centerRight))),
          ),
        ),
      ],
    );
  }

  Widget gradientDemo() {
    return SizedBox(
      height: 50,
      child: ListView.builder(
          scrollDirection: Axis.horizontal,
          itemCount: 5,
          shrinkWrap: true,
          itemBuilder: (context, index) {
            return Padding(
              padding: EdgeInsets.only(right: index == 4 ? 0 : 10),
              child: FRadio(
                width: 50,
                height: 50,
                value: index + 1,
                groupValue: groupValue_4,
                onChanged: (value) {
                  setState(() {
                    groupValue_4 = value;
                  });
                },
                gradient: LinearGradient(
                  colors: [
                    Color(0xffFEFDBB),
                    Color(0xffFFE16C),
                    Color(0xffEA9D1C),
                    Color(0xffD46307),
                  ],
                  begin: Alignment(-0.1, -0.9),
                  end: Alignment(1.0, 1.0),
                  stops: [0.0, 0.2, 0.7, 1.0],
                ),
                selectedColor: Color(0xffffc900),
                hasSpace: false,
                border: 1.5,
                child: Text(
                  "\$${5 * (index + 1)}",
                  style: TextStyle(color: Color(0xffd9d9d9), fontSize: 13),
                ),
                hoverChild: Text(
                  "\$${5 * (index + 1)}",
                  style:
                      TextStyle(color: Colors.deepOrangeAccent, fontSize: 13),
                ),
                selectedChild: Text("\$${5 * (index + 1)}",
                    style: TextStyle(
                        color: Colors.deepOrangeAccent, fontSize: 13)),
              ),
            );
          }),
    );
  }

  Widget childDemo() {
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        Column(
          children: [
            Text("EN.."),
            const SizedBox(height: 9),
            FRadio(
              width: 80,
              height: 80,
              value: 1,
              groupValue: groupValue_5,
              onChanged: (value) {
                setState(() {
                  groupValue_5 = value;
                });
              },
              child: Image.asset("assets/emoji_0.png", width: 50),
              hoverChild: Image.asset("assets/emoji_1.png", width: 50),
              selectedChild: Image.asset("assets/emoji_2.png", width: 50),
              hasSpace: false,
              toggleable: true,
              selectedColor: Color(0xffffc900),
              border: 1.5,
            ),
          ],
        ),
        const SizedBox(width: 56),
        Column(
          children: [
            Text("YA.."),
            const SizedBox(height: 9),
            FRadio(
              width: 80,
              height: 80,
              value: 2,
              groupValue: groupValue_5,
              onChanged: (value) {
                setState(() {
                  groupValue_5 = value;
                });
              },
              child: Image.asset("assets/emoji_3.png", width: 50),
              hoverChild: Image.asset("assets/emoji_4.png", width: 50),
              selectedChild: Image.asset("assets/emoji_5.png", width: 50),
              hasSpace: false,
              toggleable: true,
              selectedColor: Color(0xffffc900),
              border: 1.5,
            ),
          ],
        ),
      ],
    );
  }

  Widget spaceCornerDemo() {
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        FRadio(
          width: 100,
          height: 50,
          value: 1,
          groupValue: groupValue_3,
          onChanged: (value) {
            setState(() {
              groupValue_3 = value;
            });
          },
          selectedColor: Color(0xffffc900),
          corner: FRadioCorner(leftTopCorner: 6, leftBottomCorner: 6),
          border: 1,
          hasSpace: false,
          selectedChild:
              Text("FSuper", style: TextStyle(color: Color(0xff333333))),
          child: Text("FSuper", style: TextStyle(color: Color(0xff333333))),
        ),
        FRadio(
          width: 100,
          height: 50,
          value: 2,
          groupValue: groupValue_3,
          onChanged: (value) {
            setState(() {
              groupValue_3 = value;
            });
          },
          selectedColor: Color(0xffffc900),
          corner: FRadioCorner.all(0),
          hasSpace: false,
          border: 1,
          selectedChild:
              Text("FButton", style: TextStyle(color: Color(0xff333333))),
          child: Text("FButton", style: TextStyle(color: Color(0xff333333))),
        ),
        FRadio(
          width: 100,
          height: 50,
          value: 3,
          groupValue: groupValue_3,
          onChanged: (value) {
            setState(() {
              groupValue_3 = value;
            });
          },
          selectedColor: Color(0xffffc900),
          corner: FRadioCorner(rightTopCorner: 6, rightBottomCorner: 6),
          hasSpace: false,
          border: 1,
          selectedChild:
              Text("FRadio", style: TextStyle(color: Color(0xff333333))),
          child: Text("FRadio", style: TextStyle(color: Color(0xff333333))),
        ),
      ],
    );
  }

  Widget enableToggleableDemo() {
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        FRadio(
          value: 1,
          groupValue: groupValue_2,
          onChanged: (value) {
            setState(() {
              groupValue_2 = value;
            });
          },
          toggleable: true,
          enable: enable_1,
          selectedColor: Color(0xffffc900),
        ),
        const SizedBox(width: 5),
        Text("Android", style: TextStyle(fontSize: 16)),
        const SizedBox(width: 16),
        FRadio(
          value: 2,
          groupValue: groupValue_2,
          onChanged: (value) {
            setState(() {
              groupValue_2 = value;
            });
          },
          enable: enable_1,
          toggleable: true,
          selectedColor: Color(0xffffc900),
        ),
        const SizedBox(width: 5),
        Text("iOS", style: TextStyle(fontSize: 16)),
        const SizedBox(width: 16),
        FRadio(
          value: 3,
          groupValue: groupValue_2,
          onChanged: (value) {
            setState(() {
              groupValue_2 = value;
            });
          },
          toggleable: true,
          enable: enable_1,
          selectedColor: Color(0xffffc900),
        ),
        const SizedBox(width: 5),
        Text("Fuchsia", style: TextStyle(fontSize: 16)),
      ],
    );
  }

  Widget baseDemo() {
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        FRadio(
          value: 1,
          groupValue: groupValue_1,
          onChanged: (value) {
            setState(() {
              groupValue_1 = value;
            });
          },
        ),
        const SizedBox(width: 5),
        Text("A", style: TextStyle(fontSize: 18)),
        const SizedBox(width: 16),
        FRadio(
          value: 2,
          groupValue: groupValue_1,
          onChanged: (value) {
            setState(() {
              groupValue_1 = value;
            });
          },
        ),
        const SizedBox(width: 5),
        Text("B", style: TextStyle(fontSize: 18)),
        const SizedBox(width: 16),
        FRadio(
          value: 3,
          groupValue: groupValue_1,
          onChanged: (value) {
            setState(() {
              groupValue_1 = value;
            });
          },
        ),
        const SizedBox(width: 5),
        Text("C", style: TextStyle(fontSize: 18)),
      ],
    );
  }

  SizedBox buildMiddleMargin() {
    return const SizedBox(
      height: 26,
    );
  }

  SizedBox buildSmallMargin() {
    return const SizedBox(
      height: 18,
    );
  }

  Padding buildDesc(String desc) {
    return Padding(
        padding: const EdgeInsets.all(8),
        child: Text(
          desc,
          textAlign: TextAlign.center,
          style: TextStyle(
            color: Colors.grey,
            fontSize: 12,
          ),
        ));
  }

  Container buildTitle(String title) {
    return Container(
      alignment: Alignment.centerLeft,
      padding: EdgeInsets.all(9),
      color: Color(0xffe0e0e0).withOpacity(0.38),
      child: Text(
        title,
        style: TextStyle(color: Colors.grey),
      ),
    );
  }
}
